Command line interface

vgeorge opened the pull request:
Contributes to #120.

We are investigating a tool starter for VEDA in [this issue](https://github.com/NASA-IMPACT/veda-ui/issues/1749).

This is a work in progress and not intended to be merged here (unless we decide to do so later).

The general idea is to provide an easy way to generate a web map SPA scaffold. It makes sense to base it on `project-seed`, as we typically use it to start our projects.

I created a `/cli` folder without touching existing files. This folder contains the generator and the templates for different variants.

**What’s included:**

* CLI generator using Commander.js and Inquirer
* Base template
* Component library variants: none, Chakra UI, React USWDS
* Template merging system that combines base + selected component library

By default, the generator creates a scaffold identical to `project-seed`. But it also allows creating with different component libraries. The next step would be to offer mapping library variants, for example, MapLibre and Mapbox.

This is very exploratory work, happy to hear any feedback.

cc @aboydnw @dzole0311 @AliceR

vgeorge committed:
feat: add CLI generator for project scaffolding

vgeorge committed:
feat: refactor CLI to use base template folder instead of copying from root

vgeorge committed:
feat: enable component library variants

vgeorge committed:
chore: update CLI Readme

vgeorge committed:
chore: update deps

vgeorge committed:
fix: remove deprecated task

vgeorge committed:
feat: add --force option to overwrite output folder

vgeorge committed:
feat: add script to generate all variants

vgeorge committed:
refactor: split generator tasks in separated files for readability and maintainability

vgeorge committed:
feat: add context to app.tsx for map components and app features

vgeorge committed:
feat: enable map variants

vgeorge committed:
fix: update license year

vgeorge committed:
fix: mapbox map should use the full height

vgeorge commented on cli/templates/base/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2023 Development Seed
```
thanks @AliceR, I updated the year in another commit

vgeorge committed:
chore: update readme

vgeorge commented:
This is ready for review. I’ve made significant changes, so it’s important to check where we stand and confirm we’re heading in the right direction.

The README should now clearly cover the CLI instructions and overall approach. I recommend running `pnpm generate-all` and testing the apps in the `cli/generated` folder.

cc @dzole0311 @AliceR

vgeorge commented:
Thanks for the feedback, @dzole0311. I agree that we can expand the scaffolding capabilities of this CLI tool, but since this PR is already quite large, I suggest we track further enhancements separately.

For now, I’d like to propose merging this and using it to replace the current `project-seed` setup. As mentioned above, I believe this version is in a good state to serve as a replacement.

I'm putting this under review by the broader frontend team (especially @danielfdsilva), and I’ll demo it during our next frontend meeting.

**Next steps:**

* Review the contents in `/cli` (documentation, ability to generate a scaffold app)
* Replace the root folder with its contents
* Bump the version to `v9` and release

danielfdsilva commented on cli/src/generator/apply-component-library.ts:
```diff
@@ -0,0 +1,51 @@
+      await fs.copy(srcPath, destPath);
+    } else {
+      await fs.copy(srcPath, destPath);
+    }
```
Code is the same on the if

danielfdsilva commented on cli/src/generator/apply-map-library.ts:
```diff
@@ -0,0 +1,52 @@
+      await fs.copy(srcPath, destPath);
+    } else {
+      await fs.copy(srcPath, destPath);
+    }
```
Code is the same

danielfdsilva commented on cli/src/generator/apply-map-library.ts:
```diff
@@ -0,0 +1,52 @@
+    };
+
+    await fs.writeJson(packageJsonPath, basePackage, { spaces: 2 });
+  }
```
This code is the same as the map library and will likely be the same with another conditional library. Could be beneficial to extract into a utility

vgeorge replied:
done

danielfdsilva commented on cli/src/generator/copy-base-template-files.ts:
```diff
@@ -0,0 +1,27 @@
+      await fs.copy(srcPath, destPath);
+    } else {
+      await fs.copy(srcPath, destPath);
+    }
```
Code is the same

AliceR commented:
Thanks for the effort here @vgeorge ! This could become very useful. IMO a couple of baseline QA tools are missing:
- ~~prettier formatting (probably run via eslint, see https://prettier.io/docs/integrating-with-linters.html)~~
- ~~Stylelint for CSS linting alongside ESLint~~
- git hooks (likely husky pre-commit `npm run lint` and pre-push `npm run test:run`)
- Log accessibility violations in development mode with @axe-core/react to automatically detect and log accessibility issues to the console when running in development mode. This expands the static evaluation of JSX with eslint-plugin-jsx-a11y.
- a basic ci.yml to run lint and test on Github

We could add all these in iteration and after discussion, but I worry that if we do not take the chance now, we might never take the extra effort and add them to the generated projects later.

Edit: I looked at the wrong package.json, prettier and Stylelint are already set up!

danielfdsilva commented on cli/src/generator/update-package-json.ts:
```diff
@@ -0,0 +1,23 @@
+  if (await fs.pathExists(pkgPath)) {
+    await replaceInFile(pkgPath, {
+      '"name": *"[^"]+"': `"name": "${projectName}"`
+    });
```
Since package.json is easily updatable as json, is there a reason to use regexp replacement?

vgeorge replied:
this file is unnecessary, I moved the replacement function upstream

danielfdsilva commented on cli/templates/component-library/chakra/package.json:
```diff
@@ -0,0 +1,8 @@
+  "dependencies": {
+    "@chakra-ui/react": "^3.8.1",
+    "@emotion/react": "^11.14.0",
+    "next-themes": "^0.4.4",
```
this is not part o chakra is it? Is it being included somewhere?

vgeorge replied:
good catch, removed

vgeorge committed:
chore: update lockfile

vgeorge committed:
refactor: remove unnecessary helper function file

vgeorge committed:
refactor: remove unnecessary helper function file

vgeorge committed:
refactor: clear duplicated code

vgeorge committed:
fix: remove wrong dependencies from chakra template

vgeorge committed:
fix: remove duplicated code

vgeorge committed:
refactor: move logic to helper function

vgeorge commented:
I’ve pushed changes to address the issues noted by @danielfdsilva, this is ready for another review.

@AliceR Thanks for the feedback. I’ve created an epic ticket (#120) to track ongoing work on the CLI and to capture your proposed changes there. Since this PR already introduces a significant number of changes, I believe it makes sense to merge it into the develop branch and track the remaining work in the epic ticket.

vgeorge committed:
fix: clear root folder contents, keep a few files

vgeorge committed:
refactor: move CLI files to root directory

vgeorge committed:
merge: resolve conflicts with develop branch

AliceR commented on templates/base/eslint.config.mjs:
```diff
@@ -25,6 +25,7 @@ export default [
       'prefer-promise-reject-errors': 0,
       // 'import/order': 2,
       'react/button-has-type': 2,
+      'react/react-in-jsx-scope': 0,
```
Is this a good moment to revisit whether we need all of these custom rules (https://github.com/developmentseed/project-seed/pull/118/commits/ef17ccf7548bf579e7fc4d735335523ceec4285b#diff-8bea3dd751e9c7b3360f24b10c80108108cdf08b37d1f149a37d9414c74a723dR21-R60) ? I wonder if the config could be simplified by removing those custom rules? I think it could be a good moment to re-evaluate now, since we're introducing a new tool.

vgeorge replied:
removed

vgeorge committed:
fix: remove custom rules

vgeorge committed:
fix: Add missing dependency on chakra template

vgeorge commented:
@AliceR I removed the ESLint custom rule. I also added a missing dependency of the chakra template.

AliceR committed:
fix: Use different template path for dev vs. prod mode

- Add getTemplatePath utility function to handle path differences between
  development (src/generator) and production (dist/) environments
- Replace hardcoded path.resolve calls with centralized function across
  all generator modules (index.ts, apply-component-library.ts, apply-map-library.ts)
- Fix template path resolution issues that caused ENOENT errors when
  running CLI in different modes
- Ensure consistent behavior between `pnpm start` (dev) and `node dist/index.js` (prod)

AliceR commented:
@vgeorge I noticed an issue with the path resolution in dev vs. prod. #121 should fix it!

vgeorge committed:
Merge pull request #121 from developmentseed/fix-template-path

fix: Use different template path for dev vs. prod mode

AliceR reviewed (approved):
IMO this is solid enough to merge to develop! 👏  We can iterate from there, or leave cli as base for further changes, I don't mind the difference :)

vgeorge commented:
@AliceR lets continue from develop, thanks for the changes 🙂
